src/v/cloud_storage/topic_recovery_service.cc

```diff
@@ -0,0 +1,594 @@
+
+    auto manifests = co_await filter_existing_topics(
+      bucket_contents, request, model::ns{"kafka"});
+
+    if (manifests.empty()) {
+        vlog(cst_log.info, "exiting recovery, no topics to create");
+        co_await propagate_state(state::inactive);
+        co_return outcome::success();
+    }
+
+    vlog(cst_log.info, "found {} topics to create", manifests.size());
+    for (const auto& manifest : manifests) {
+        vlog(cst_log.debug, "topic manifest: {}", manifest.get_manifest_path());
+    }
+    _recovery_status.clear();
+
+    auto clear_fib = make_rtc(_as, _config);
+    co_await clear_recovery_results(
+      _remote.local(), _config.bucket, clear_fib, std::nullopt);
+    _downloaded_manifests.emplace(manifests);
+
+    populate_recovery_status();
+
+    co_await propagate_state(state::creating_topics);
+    vlog(cst_log.debug, "creating topics");
+
+    result<void, recovery_error_ctx> result = outcome::success();
+    try {
+        auto create_topics_results = co_await create_topics(request);
+        for (const auto& r : create_topics_results) {
+            if (r.ec != cluster::errc::success) {
+                vlog(cst_log.warn, "topic creation failed: {}", r);
+                _recovery_status.erase(r.tp_ns);
+            } else {
+                vlog(
+                  cst_log.debug,
+                  "topic created: {}, expected downloads: {}",
+                  r,
+                  _recovery_status[r.tp_ns].pending_downloads);
+            }
+        }
+
+        co_await propagate_state(state::recovering_data);
+        start_download_bg_tracker();
+    } catch (const ss::timed_out_error& err) {
+        result = recovery_error_ctx::make(
+          fmt::format("failed to create topics: {}", err),
+          recovery_error_code::error_creating_topics);
+    }
+
+    if (result.has_error()) {
+        co_await propagate_state(state::inactive);
+    }
+    co_return result;
+}
+
+static cluster::topic_configuration make_topic_config(
+  const cloud_storage::topic_manifest& tm, const recovery_request& request) {
+    auto topic_config = tm.get_topic_config();
+    if (!topic_config.has_value()) {
+        vlog(
+          cst_log.warn,
+          "skipping topic creation for {}, missing values in manifest",
+          tm.get_manifest_path());
+    }
+
+    cluster::topic_configuration topic_to_create_cfg(
+      topic_config->tp_ns.ns,
+      topic_config->tp_ns.tp,
+      topic_config->partition_count,
+      topic_config->replication_factor);
+
+    auto& topic_properties = topic_to_create_cfg.properties;
+    auto manifest_props = topic_config->properties;
+
+    topic_properties.compression = manifest_props.compression;
+    topic_properties.cleanup_policy_bitflags
+      = manifest_props.cleanup_policy_bitflags;
+    topic_properties.compaction_strategy = manifest_props.compaction_strategy;
+
+    topic_properties.retention_duration = tristate<std::chrono::milliseconds>{
```

Comment:
Why set retention_duration, rather than just `retention_local_target_ms`?

(I vaguely recall that the existing topic recovery code might have been using the wrong setting, if so we should fix that)

Reply:
changed to only set the local value. also see https://github.com/redpanda-data/redpanda/pull/7705#discussion_r1080826278